lib/src/revset.rs

```diff
@@ -638,9 +641,19 @@ static BUILTIN_FUNCTION_MAP: Lazy<HashMap<&'static str, RevsetFunction>> = Lazy:
         } else {
             StringPattern::everything()
         };
+        let remote_ref_state = if let Some(tracked_arg) = tracked_opt_arg {
+            if revset_parser::expect_bool(tracked_arg)? {
+                Some(RemoteRefState::Tracking)
+            } else {
+                Some(RemoteRefState::New)
+            }
+        } else {
+            None
```

Comment:
I think the default should also be explicitly specifiable by user, by using something like `tracked|untracked|any`? Separate `tracked_remote_branches()`/`untracked_remote_branches()` functions seem also good.

Thoughts?

Reply:
I agree that having a way to explicitly specify the default value would be good. Maybe the argument name would be `state` then (e.g. `remote_branches(state=untracked)`)?

I think that having `tracked_remote_branches()` and `untracked_remote_branches()` would work as well, and it might be simpler due to not having to parse `tracked`, `untracked`, and `any`. It might also be more discoverable due to function name suggestions.

On the other hand, tracked and untracked remote branches don't seem like fundamentally different types of branches to me (at least not as different as local branches and remote branches are), so I'm not sure that they're different enough for it to make sense to have them be different revsets. I also think it would result in more duplication, since we'd probably want to support filtering by branch name and remote name for all 3 revsets.

I don't have a strong feeling either way. Which do you think is better?

Reply:
I don't have strong preference either. `remote_branches(state=untracked)` looks good to me. If a typical use case is `remote_branches(state=untracked)` (without name pattern), `untracked_remote_branches()` might be better. The downside is that it's uneasy to extend if we have more search options.

@ilyagr any comments?

In any case, the underlying `RevsetExpression` enum can shared. (i.e. `{ branch, remote, state: Option<State> }`.)

Reply:
> If a typical use case is `remote_branches(state=untracked)` (without name pattern), `untracked_remote_branches()` might be better. 

My guess would be that it would usually be used without any names (e.g. marking all untracked branches as immutable), but sometimes it would be useful to filter by branch or remote.

> The downside is that it's uneasy to extend if we have more search options.

Yeah, in general it seems impractical to add new revsets for each boolean filter option, since with 2 filters we'd need 9 revsets to cover every combination. I can't think of any more filter options that would apply to remote branches though, so maybe that wouldn't be an issue.

I updated it to use the `state` argument version for now since it was a small change, but I'll wait to hear what @ilyagr thinks.

Reply:
I think `untracked_remote_branches()` and `tracked_remote_branches()` is pretty safe for now.

Another option we could perhaps consider is `remote_branches(tracked=true/false/any)`.

I like `state=tracked/untracked` the least, for two reasons:

- I think generic-sounding keywords like `state=` are annoying and difficult to remember.
- We might want to add "branches(tracked_by_remotes=....)" or something to that effect. That would not work with "state".

Though it would also do for now.

-------

I wonder if we eventually need a general solution for this, a "branchset language" of some sort.
We could have a general `branchexpr()` function and have `remote_branches(x) = branchexpr( remote() & x )`.

Reply:
> Another option we could perhaps consider is `remote_branches(tracked=true/false/any)`.

I personally find it a bit strange to have `any` for a boolean option, since most programming languages don't use a three-valued logic, so I think I'd prefer `untracked_remote_branches()` and `tracked_remote_branches()`.

> I like `state=tracked/untracked` the least, for two reasons:
>
>   * I think generic-sounding keywords like `state=` are annoying and difficult to remember.
>
>   * We might want to add `branches(tracked_by_remotes="glob:work-*")` or something to that effect for local branches. That would not work with "state".

That makes sense to me.

> I wonder if we eventually need a general solution for this, a "branchset language" of some sort. We could have a general `branchexpr()` function and have `remote_branches(x) = branchexpr( remote() & x )`.

I had actually been thinking about that! If we were going to add more filters, that would probably be nicer than just adding them as arguments or separate revsets.

Reply:
Let's settle on `remote_branches()`/`tracked_remote_branches()`/`untracked_remote_branches()` then. I don't like tri-state boolean, and I agree `state=` is a bit vague word.

Implementation-wise, inner arguments parsing function can be extracted.